### PR TITLE
Remove Transfer Overflow Checks

### PR DESCRIPTION
**Pull request by hexonaut**
Overflow checks are not necessary as the amount of DAI being transferred is bounded by the user balances which is well below 2^256 during normal operation. In the event the `mint` is compromised all bets are off anyways.

This still needs discussion as to whether to `to` should be also checked to not equal the token address.

### CODE REVIEW

File: contracts/l2/dai.sol [hunk 1]
```diff
@@ -96,17 +93,19 @@ contract Dai {
...
     balanceOf[msg.sender] = balance - value;
-    balanceOf[to] = add(balanceOf[to], value);
+    balanceOf[to] = balanceOf[to] + value;
 
     emit Transfer(msg.sender, to, value);
 
     return true;
   }
   function transferFrom(address from, address to, uint256 value) external returns (bool) {
+    require(to != address(0), "Dai/invalid-address");
```
Comment (naszam) [inline contracts/l2/dai.sol:108]: I would add another require for `from` as well.
`require(from != address(0), "Dai/invalid-address");`

File: contracts/l2/dai.sol [hunk 2]
```diff
@@ -96,17 +93,19 @@ contract Dai {
 
   // --- ERC20 Mutations ---
   function transfer(address to, uint256 value) external returns (bool) {
+    require(to != address(0), "Dai/invalid-address");
     uint256 balance = balanceOf[msg.sender];
     require(balance >= value, "Dai/insufficient-balance");
 
     balanceOf[msg.sender] = balance - value;
-    balanceOf[to] = add(balanceOf[to], value);
+    balanceOf[to] = balanceOf[to] + value;
```
Comment (naszam) [inline contracts/l2/dai.sol:101]: I would do like:
`balanceOf[to] += value;`

File: contracts/l2/dai.sol [hunk 3]
```diff
@@ -120,7 +119,7 @@ contract Dai {
     }
 
     balanceOf[from] = balance - value;
-    balanceOf[to] = add(balanceOf[to], value);
+    balanceOf[to] = balanceOf[to] + value;
```
Comment (naszam) [inline contracts/l2/dai.sol:122]: same comment as above

File: contracts/l2/dai.sol (hunk 1 above)
Comment (hexonaut) [reply contracts/l2/dai.sol:108]: Setting `from` to `address(0)` will fail anyways as there is no allowance. We don't need to add this explicit check.

File: contracts/l2/dai.sol [hunk 4]
```diff
@@ -72,9 +72,6 @@ contract Dai {
     assembly {chainId := chainid()}
     deploymentChainId = chainId;
     _DOMAIN_SEPARATOR = _calculateDomainSeparator(chainId);
-
-    // Set addresses which disallow transfer
-    balanceOf[address(this)] = balanceOf[address(0)] = type(uint256).max;
```
Comment (naszam) [inline contracts/l2/dai.sol:77]: has been reached consensus on this?

Comment (hexonaut) [reply contracts/l2/dai.sol:77]: This whole PR is removing the overflow check. So we first need to decide if this PR is okay, and if so whether we want to re-add the disallow send to contract check. I'm going to be grabbing some numbers first to aid with this decision.

**Review by krzkaczor — COMMENTED** [review]
With this change, we lose the check against sending DAI to the contract itself. I think we already discussed that there could be a "Customer Support Core Unit" responsible for retrieving lost DAI from the contract. Since gov relayer is an admin of this contract it's possible to create a spell that would mint missing tokens back to the owners. 

Unfortunately, this would break the accounting as supply would inflate. Should we consider making a change in `burn()` to allow burning stuck tokens (`0` and `this` addr)? Maybe then checks against 0 address are also not needed?

**Comment by hexonaut** [comment]
You are correct. I don't think we need anything fancy beyond just the check for `address(this)` if we decide to go that route. I am still against adding this because I believe it belongs in the UI. I have some data we can go over to see if this is a worthwhile thing to add.

**Comment by hexonaut** [comment]
As per the meeting today we are going to keep the check for sending dai to the token address. This is good for final review / merge.

File: contracts/l2/dai.sol [hunk 5]
```diff
@@ -96,17 +93,19 @@ contract Dai {
 
   // --- ERC20 Mutations ---
   function transfer(address to, uint256 value) external returns (bool) {
+    require(to != address(0) && to != address(this), "Dai/invalid-address");
     uint256 balance = balanceOf[msg.sender];
     require(balance >= value, "Dai/insufficient-balance");
 
     balanceOf[msg.sender] = balance - value;
-    balanceOf[to] = add(balanceOf[to], value);
+    balanceOf[to] += value;
```
Comment (naszam) [inline contracts/l2/dai.sol:101]: just to double check, where is the overflow checked on this? (z >= x)

File: contracts/l2/dai.sol [hunk 6]
```diff
@@ -120,7 +119,7 @@ contract Dai {
     }
 
     balanceOf[from] = balance - value;
-    balanceOf[to] = add(balanceOf[to], value);
+    balanceOf[to] += value;
```
Comment (naszam) [inline contracts/l2/dai.sol:122]: same as above

File: test/l2/dai.ts [hunk 7]
```diff
@@ -84,10 +84,21 @@ describe('Dai', () => {
         ).to.be.revertedWith('')
       })
 
+      it('should not transfer to dai address', async () => {
+        await expect(dai.connect(signers.user1).transfer(dai.address, 1)).to.be.revertedWith('')
+        await expect(dai.connect(signers.user1).transferFrom(signers.user1.address, dai.address, 1)).to.be.revertedWith(
```
Comment (naszam) [inline test/l2/dai.ts]: just a note, `revertedWith` expects a message error to be passed as parameter. if that's not the case `reverted` is the preferred way.

File: contracts/l2/dai.sol (hunk 5 above)
Comment (naszam) [reply contracts/l2/dai.sol:101]: I'll run a fuzzing campaign to check on this, probably improving the current echidna tests to check for overflows.

**Review by krzkaczor — APPROVED** [review]

**Pull request merged**